README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> ElasticSearch components for Dart

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The Elasticsearch module contains logging components with data storage on the Elasticsearch server.

The module contains the following packages:
- **Build** - contains a factory for the construction of components
- **Log** - Logging components


<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [Logging](https://www.pipservices.org/recipies/active-logic)
* [Virtual memory configuration](https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html)
* [API Reference](https://pub.dev/documentation/pip_services3_elasticsearch/latest/pip_services3_elasticsearch/pip_services3_elasticsearch-library.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)


## Use

Add this to your package's pubspec.yaml file:
```yaml
dependencies:
  pip_services3_elasticsearch: version
```

Now you can install package from the command line:
```bash
pub get
```

Microservice components shall perform logging usual way using CompositeLogger component.
The CompositeLogger will get ElasticSearchLogger from references and will redirect log messages
there among other destinations.

```dart
class MyComponent implements IConfigurable, IReferenceable {
  CompositeLogger _logger = new CompositeLogger();
  
  configure(ConfigParams config) {
    this._logger.configure(config);
  }
  
  setReferences(IReferences refs) {
    this._logger.setReferences(refs);
  }
  
  myMethod(String? correlationId, param1) {
    this._logger.trace(correlationId, "Executed method mycomponent.mymethod");
    ....
  }
}
```

Configuration for your microservice that includes ElasticSearch logger may look the following way.

```yaml
...
{{#if ELASTICSEARCH_ENABLED}}
- descriptor: pip-services:logger:elasticsearch:default:1.0
  connection:
    uri: {{{ELASTICSEARCG_SERVICE_URI}}}
    host: {{{ELASTICSEARCH_SERVICE_HOST}}}{{#unless ELASTICSEARCH_SERVICE_HOST}}localhost{{/unless}}
    port: {{ELASTICSEARCG_SERVICE_PORT}}{{#unless ELASTICSEARCH_SERVICE_PORT}}9200{{/unless}}\ 
{{/if}}
...
```

## Develop

For development you shall install the following prerequisites:
* Dart SDK 2
* Visual Studio Code or another IDE of your choice
* Docker

Install dependencies:
```bash
pub get
```

Run automated tests:
```bash
pub run test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by 
- **Sergey Seroukhov**
- **Levichev Dmitry**

The documentation is written by:
- **Mark Makarychev**
- **Levichev Dmitry**

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> ElasticSearch components for Dart Changelog

## 1.0.0

- Initial version, created by Sergey Seroukhov and Dmitry Levichev

## 1.1.0

- Migrated to null safety
- Added supports ElasticSearch 7.x "typeless" indexes as well as 6.x "typed" indexes.


test/fixtures/LoggerFixture.dart
import 'dart:async';

import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:test/test.dart';
import 'package:pip_services3_components/pip_services3_components.dart';

class LoggerFixture {
  late CachedLogger _logger;

  LoggerFixture(CachedLogger logger) {
    _logger = logger;
  }

  void testLogLevel() {
    expect(_logger.getLevel().index >= LogLevel.None.index, isTrue);
    expect(_logger.getLevel().index <= LogLevel.Trace.index, isTrue);
  }

  Future testSimpleLogging() async {
    _logger.setLevel(LogLevel.Trace);

    _logger.fatal(null, null, 'Fatal error message');
    _logger.error(null, null, 'Error message');
    _logger.warn(null, 'Warning message');
    _logger.info(null, 'Information message');
    _logger.debug(null, 'Debug message');
    _logger.trace(null, 'Trace message');

    _logger.dump();
    await Future.delayed(Duration(milliseconds: 1000));
  }

  Future testErrorLogging() async {
    try {
      // Raise an exception
      throw Exception('Test error');
    } catch (err) {
      var ex = ApplicationException().wrap(err);
      _logger.fatal('123', ex, 'Fatal error');
      _logger.error('123', ex, 'Recoverable error');

      expect(ex, isNotNull);
    }

    _logger.dump();
    await Future.delayed(Duration(milliseconds: 1000));
  }
}


test/log/ElasticSearchLogger.test.dart
// import { ConfigParams } from 'pip-services3-commons-node';

// import { ElasticSearchLogger } from '../../src/log/ElasticSearchLogger';
// import { LoggerFixture } from '../fixtures/LoggerFixture';

// let assert = require('chai').assert;
// let async = require('async');

// suite('ElasticSearchLogger', () => {
//     let _logger: ElasticSearchLogger;
//     let _fixture: LoggerFixture;

//     setup((done) => {
//         let host = process.env['ELASTICSEARCH_SERVICE_HOST'] || 'localhost';
//         let port = process.env['ELASTICSEARCH_SERVICE_PORT'] || 9200;
//         let dateFormat: string = "YYYYMMDD";

//         _logger = new ElasticSearchLogger();
//         _fixture = new LoggerFixture(_logger);

//         let config = ConfigParams.fromTuples(
//             'source', 'test',
//             'index', 'log',
//             'daily', true,
//             "date_format", dateFormat,
//             'connection.host', host,
//             'connection.port', port
//         );
//         _logger.configure(config);

//         _logger.open(null, (err) => {
//             done(err);
//         });
//     });

//     teardown((done) => {
//         _logger.close(null, done);
//     });

//     test('Log Level', () => {
//         _fixture.testLogLevel();
//     });

//     test('Simple Logging', (done) => {
//         _fixture.testSimpleLogging(done);
//     });

//     test('Error Logging', (done) => {
//         _fixture.testErrorLogging(done);
//     });

//     
//     /// We test to ensure that the date pattern does not effect the opening of the elasticsearch component
//      
//     test('Date Pattern Testing - YYYY.MM.DD', (done) => {

//         let host = process.env['ELASTICSEARCH_SERVICE_HOST'] || 'localhost';
//         let port = process.env['ELASTICSEARCH_SERVICE_PORT'] || 9200;

//         let logger = new ElasticSearchLogger();
//         let dateFormat: string = "YYYY.MM.DD";

//         let config = ConfigParams.fromTuples(
//             'source', 'test',
//             'index', 'log',
//             'daily', true,
//             "date_format", dateFormat,
//             'connection.host', host,
//             'connection.port', port
//         );
//         logger.configure(config);

//         logger.open(null, (err) => {
//             if (err) {
//                 done(err);
//             }

//             // Since the currentIndex property is private, we will just check for an open connection
//             assert.isTrue(logger.isOpen());
//             logger.close(null, done);
//         });
//     });

//     
//     /// We test to ensure that the date pattern does not effect the opening of the elasticsearch component
//      
//     test('Date Pattern Testing - YYYY.M.DD', (done) => {

//         let host = process.env['ELASTICSEARCH_SERVICE_HOST'] || 'localhost';
//         let port = process.env['ELASTICSEARCH_SERVICE_PORT'] || 9200;

//         let logger = new ElasticSearchLogger();
//         let dateFormat: string = "YYYY.M.DD";

//         let config = ConfigParams.fromTuples(
//             'source', 'test',
//             'index', 'log',
//             'daily', true,
//             "date_format", dateFormat,
//             'connection.host', host,
//             'connection.port', port
//         );
//         logger.configure(config);

//         logger.open(null, (err) => {
//             if (err) {
//                 done(err);
//             }

//             // Since the currentIndex property is private, we will just check for an open connection
//             assert.isTrue(logger.isOpen());
//             logger.close(null, done);
//         });
//     });

// });

test/log/ElasticSearchLogger_test.dart
import 'dart:io';
import 'package:test/test.dart';
import 'package:pip_services3_commons/pip_services3_commons.dart';

import 'package:pip_services3_elasticsearch/pip_services3_elasticsearch.dart';
import '../fixtures/LoggerFixture.dart';

void main() {
  group('ElasticSearchLogger', () {
    late ElasticSearchLogger _logger;
    late LoggerFixture _fixture;

    setUp(() async {
      var host =
          Platform.environment['ELASTICSEARCH_SERVICE_HOST'] ?? 'localhost';
      var port = Platform.environment['ELASTICSEARCH_SERVICE_PORT'] ?? 9200;
      var dateFormat = 'yyyyMMdd';

      _logger = ElasticSearchLogger();
      _fixture = LoggerFixture(_logger);

      var config = ConfigParams.fromTuples([
        'source',
        'test',
        'index',
        'log',
        'daily',
        true,
        'date_format',
        dateFormat,
        'connection.host',
        host,
        'connection.port',
        port
      ]);
      _logger.configure(config);

      await _logger.open(null);
    });

    tearDown(() async {
      await _logger.close(null);
    });

    test('Log Level', () {
      _fixture.testLogLevel();
    });

    test('Simple Logging', () async {
      await _fixture.testSimpleLogging();
    });

    test('Error Logging', () async {
      await _fixture.testErrorLogging();
    });

    /// We test to ensure that the date pattern does not effect the opening of the elasticsearch component

    test('Date Pattern Testing - yyyy.MM.dd', () async {
      var host =
          Platform.environment['ELASTICSEARCH_SERVICE_HOST'] ?? 'localhost';
      var port = Platform.environment['ELASTICSEARCH_SERVICE_PORT'] ?? 9200;

      var logger = ElasticSearchLogger();
      var dateFormat = 'yyyy.MM.dd';

      var config = ConfigParams.fromTuples([
        'source',
        'test',
        'index',
        'log',
        'daily',
        true,
        'date_format',
        dateFormat,
        'connection.host',
        host,
        'connection.port',
        port
      ]);
      logger.configure(config);

      await logger.open(null);

      // Since the currentIndex property is private, we will just check for an open connection
      expect(logger.isOpen(), isTrue);
      await logger.close(null);
    });

    /// We test to ensure that the date pattern does not effect the opening of the elasticsearch component

    test('Date Pattern Testing - yyyy.M.dd', () async {
      var host =
          Platform.environment['ELASTICSEARCH_SERVICE_HOST'] ?? 'localhost';
      var port = Platform.environment['ELASTICSEARCH_SERVICE_PORT'] ?? 9200;

      var logger = ElasticSearchLogger();
      var dateFormat = 'yyyy.M.dd';

      var config = ConfigParams.fromTuples([
        'source',
        'test',
        'index',
        'log',
        'daily',
        true,
        'date_format',
        dateFormat,
        'connection.host',
        host,
        'connection.port',
        port
      ]);
      logger.configure(config);

      await logger.open(null);

      // Since the currentIndex property is private, we will just check for an open connection
      expect(logger.isOpen(), isTrue);
      await logger.close(null);
    });
  });
}


lib/src/log/ElasticSearchLogger.dart
import 'dart:async';

import 'package:elastic_client/elastic_client.dart';
import 'package:intl/intl.dart';

import 'package:elastic_client/elastic_client.dart' as elastic;

import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_rpc/pip_services3_rpc.dart';
import 'package:pip_services3_components/pip_services3_components.dart';

/// Logger that dumps execution logs to ElasticSearch service.
///
/// ElasticSearch is a popular search index. It is often used
/// to store and index execution logs by itself or as a part of
/// ELK (ElasticSearch - Logstash - Kibana) stack.
///
/// Authentication is not supported in this version.
///
/// ### Configuration parameters ###
///
/// - [level]:             maximum log level to capture
/// - [source]:            source (context) name
/// - [connection(s)]:
///     - [discovery_key]:         (optional) a key to retrieve the connection from [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html)
///     - [protocol]:              connection protocol: http or https
///     - [host]:                  host name or IP address
///     - [port]:                  port number
///     - [uri]:                   resource URI or connection string with all parameters in it
/// - [options]:
///     - [interval]:        interval in milliseconds to save log messages (default: 10 seconds)
///     - [max_cache_size]:  maximum number of messages stored in this cache (default: 100)
///     - [index]:           ElasticSearch index name (default: 'log')
///     - [date_format]      The date format to use when creating the index name. Eg. log-yyyyMMdd (default: 'yyyyMMdd'). See [DateFormat](https://pub.dev/documentation/intl/latest/intl/DateFormat-class.html)
///     - [daily]:           true to create a new index every day by adding date suffix to the index
///                        name (default: false)
///     - [reconnect]:       reconnect timeout in milliseconds (default: 60 sec)
///     - [timeout]:         invocation timeout in milliseconds (default: 30 sec)
///     - [max_retries]:     maximum number of retries (default: 3)
///     - [index_message]:   true to enable indexing for message object (default: false)
///     - [include_type_name]: Will create using a "typed" index compatible with ElasticSearch 6.x (default: false)
///
/// ### References ###
///
/// - \*:context-info:\*:\*:1.0      (optional) [ContextInfo](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/ContextInfo-class.html) to detect the context id and specify counters source
/// - \*:discovery:\*:\*:1.0         (optional) [IDiscovery](https://pub.dev/documentation/pip_services3_components/latest/pip_services3_components/IDiscovery-class.html) services to resolve connection
///
/// ### Example ###
///
///     var logger = ElasticSearchLogger();
///     logger.configure(ConfigParams.fromTuples([
///         'connection.protocol', 'http',
///         'connection.host', 'localhost',
///         'connection.port', 9200
///     ]));
///
///     await logger.open('123')
///         ...
///
///     logger.error('123', ex, 'Error occured: %s', ex.message);
///     logger.debug('123', 'Everything is OK.');

class ElasticSearchLogger extends CachedLogger
    implements IReferenceable, IOpenable {
  final _connectionResolver = HttpConnectionResolver();

  Timer? _timer;
  String _index = 'log';
  String _dateFormat = 'yyyyMMdd';
  bool _dailyIndex = false;
  String? _currentIndex;
  int _reconnect = 60000;
  int _timeout = 30000;
  int _maxRetries = 3;
  bool _indexMessage = false;
  bool _includeTypeName = false;

  elastic.Client? _client;
  HttpTransport? _transport;

  /// Creates a new instance of the logger.
  ElasticSearchLogger() : super();

  /// Configures component by passing configuration parameters.
  ///
  /// -  [config]    configuration parameters to be set.
  @override
  void configure(ConfigParams config) {
    super.configure(config);

    _connectionResolver.configure(config);

    _index = config.getAsStringWithDefault('index', _index);
    _dateFormat = config.getAsStringWithDefault('date_format', _dateFormat);
    _dailyIndex = config.getAsBooleanWithDefault('daily', _dailyIndex);
    _reconnect =
        config.getAsIntegerWithDefault('options.reconnect', _reconnect);
    _timeout = config.getAsIntegerWithDefault('options.timeout', _timeout);
    _maxRetries =
        config.getAsIntegerWithDefault('options.max_retries', _maxRetries);
    _indexMessage =
        config.getAsBooleanWithDefault('options.index_message', _indexMessage);
    _includeTypeName = config.getAsBooleanWithDefault(
        'options.include_type_name', _includeTypeName);
  }

  /// Sets references to dependent components.
  ///
  /// -  [references] 	references to locate the component dependencies.
  @override
  void setReferences(IReferences references) {
    super.setReferences(references);
    _connectionResolver.setReferences(references);
  }

  /// Checks if the component is opened.
  ///
  /// Return true if the component has been opened and false otherwise.
  @override
  bool isOpen() {
    return _timer != null;
  }

  /// Opens the component.
  ///
  /// -  [correlationId] 	(optional) transaction id to trace execution through call chain.
  /// Return 			Future that receives null no errors occured.
  /// Throws error
  @override
  Future open(String? correlationId) async {
    if (isOpen()) {
      return null;
    }

    var connection = await _connectionResolver.resolve(correlationId);
    if (connection == null) {
      throw ConfigException(
          correlationId, 'NO_CONNECTION', 'Connection is not configured');
    }

    var uri = connection.getAsString('uri');

    // var options = {
    //     host: uri,
    //     requestTimeout: this._timeout,
    //     deadTimeout: this._reconnect,
    //     maxRetries: this._maxRetries
    // };

    _transport = HttpTransport(url: Uri.parse(uri));
    _client = elastic.Client(_transport!);

    await _createIndexIfNeeded(correlationId, true);
    _timer = Timer.periodic(Duration(milliseconds: interval), (tm) {
      dump();
    });
  }

  /// Closes component and frees used resources.
  ///
  /// -  [correlationId] 	(optional) transaction id to trace execution through call chain.
  /// Return 			Future that receives null no errors occured.
  /// Throws error
  @override
  Future close(String? correlationId) async {
    await save(cache);
    if (_timer != null) {
      _timer!.cancel();
    }

    cache = [];
    _timer = null;
    _client = null;
    await _transport?.close();
    _transport = null;
  }

  String _getCurrentIndex() {
    if (!_dailyIndex) return _index;

    return _index +
        '-' +
        DateFormat(_dateFormat).format(DateTime.now().toUtc());
  }

  Future _createIndexIfNeeded(String? correlationId, bool force) async {
    var newIndex = _getCurrentIndex();
    if (!force && _currentIndex == newIndex) {
      return null;
    }

    _currentIndex = newIndex;
    var exists = await _client!.indexExists(index: _currentIndex!);

    if (exists) {
      return null;
    }

    await _client!.updateIndex(index: _currentIndex!, content: {
      'settings': {'number_of_shards': 1},
      'mappings': _getIndexSchema()
    });
  }

  /// Returns the schema of the log message
  /// - [include_type_name] A flag that indicates whether the schema should follow the pre-ES 6.x convention
  Map<String, dynamic> _getIndexSchema() {
    var schema = {
      'properties': {
        'time': {'type': 'date', 'index': true},
        'source': {'type': 'keyword', 'index': true},
        'level': {'type': 'keyword', 'index': true},
        'correlation_id': {'type': 'text', 'index': true},
        'error': {
          'type': 'object',
          'properties': {
            'type': {'type': 'keyword', 'index': true},
            'category': {'type': 'keyword', 'index': true},
            'status': {'type': 'integer', 'index': false},
            'code': {'type': 'keyword', 'index': true},
            'message': {'type': 'text', 'index': false},
            'details': {'type': 'object'},
            'correlation_id': {'type': 'text', 'index': false},
            'cause': {'type': 'text', 'index': false},
            'stack_trace': {'type': 'text', 'index': false}
          }
        },
        'message': {'type': 'text', 'index': _indexMessage}
      }
    };
    if (_includeTypeName) {
      return {'log_message': schema};
    } else {
      return schema;
    }
  }

  /// Saves log messages from the cache.
  ///
  /// -  [messages]  a list with log messages
  /// Return  Future that receives null for success.
  /// Throws error
  @override
  Future save(List<LogMessage> messages) async {
    if (!isOpen() && messages.isEmpty) {
      return null;
    }

    await _createIndexIfNeeded('elasticsearch_logger', false);

    var logItem = getLogItem();

    var bulk = <Doc>[];
    for (var message in messages) {
      var doc = Doc(logItem['_id'], message.toJson(),
          index: logItem['_index'], type: logItem['_type']);
      bulk.add(doc);
    }

    var compleate = await _client!.bulk(updateDocs: bulk);
    if (!compleate) {
      throw ApplicationException('Logger', 'elasticsearch_logger', 'SAVE_ERROR',
          'Can\'t save log messages to Elasticsearch server!');
    }
  }

  Map<String, dynamic> getLogItem() {
    return _includeTypeName
        ? {
            '_index': _currentIndex,
            '_type': 'log_message',
            '_id': IdGenerator.nextLong()
          } // ElasticSearch 6.x
        : {
            '_index': _currentIndex,
            '_id': IdGenerator.nextLong()
          }; // ElasticSearch 7.x
  }
}


example/main.dart
import 'dart:io';
import 'package:pip_services3_commons/pip_services3_commons.dart';
import 'package:pip_services3_components/pip_services3_components.dart';
import 'package:pip_services3_elasticsearch/pip_services3_elasticsearch.dart';

void main() async {
  var host = Platform.environment['ELASTICSEARCH_SERVICE_HOST'] ?? 'localhost';
  var port = Platform.environment['ELASTICSEARCH_SERVICE_PORT'] ?? 9200;
  var dateFormat = 'yyyyMMdd';

  var _logger = ElasticSearchLogger();

  var config = ConfigParams.fromTuples([
    'source',
    'test',
    'index',
    'log',
    'daily',
    true,
    'date_format',
    dateFormat,
    'connection.host',
    host,
    'connection.port',
    port
  ]);
  _logger.configure(config);

  await _logger.open(null);

  _logger.setLevel(LogLevel.Trace);
  _logger.fatal(null, null, 'Fatal error message');
  _logger.error(null, null, 'Error message');
  _logger.warn(null, 'Warning message');
  _logger.info(null, 'Information message');
  _logger.debug(null, 'Debug message');
  _logger.trace(null, 'Trace message');

  try {
    // Raise an exception
    throw Exception('Test error');
  } catch (err) {
    var ex = ApplicationException().wrap(err);
    _logger.fatal('123', ex, 'Fatal error');
    _logger.error('123', ex, 'Recoverable error');
  }

  _logger.dump();

  await _logger.close(null);
}
